### PR TITLE
Test(pyavd): Add more tests of schema tooling

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
In preparation of a major schema tooling refactor (move to Rust), this adds full coverage of the core conversion and validation tooling within the pytests.

Only exception is schemas with only dynamic_keys, but that would never happen as dynamic keys would need to point to some regular keys to be useful.

### CODE REVIEW

File: python-avd/tests/pyavd/schema/test_avddataconverter.py
Comment (carl-baillargeon) [inline python-avd/tests/pyavd/schema/test_avddataconverter.py:1]: `convert_keys`, `convert_dynamic_keys` and `convert_items` have early returns if the data type doesn't match the schema expectation. Do we need to test this?

File: python-avd/tests/pyavd/schema/test_avddataconverter.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+def test_convert_types_for_list_items_invalid() -> None:
+    schema = {"type": "list", "items": {"type": "int", "convert_types": ["str"]}}
+    avdschema = AvdSchema(schema=schema)
+    data = ["not an int", 456]
+    warnings = list(avdschema.convert(data))
+    assert len(warnings) == 0
+    assert data == ["not an int", 456]
+
+
+def test_to_lower_case_for_list_items() -> None:
```
Comment (carl-baillargeon) [inline python-avd/tests/pyavd/schema/test_avddataconverter.py:28]: Shoud we also test converting the values of dictionary keys to lowercase?

File: python-avd/tests/pyavd/schema/test_avddataconverter.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        },
+    }
+
+    avdschema = AvdSchema(schema=schema)
+    data = {"a": True, "b": True}
+    warnings = list(avdschema.convert(data))
+    assert len(warnings) == 0
+
+
+def test_deprecation_removed() -> None:
```
Comment (carl-baillargeon) [inline python-avd/tests/pyavd/schema/test_avddataconverter.py:115]: Shoud we test `remove_after_date`?

File: python-avd/tests/pyavd/schema/test_avddataconverter.py (hunk 1 above)
Comment (carl-baillargeon) [reply python-avd/tests/pyavd/schema/test_avddataconverter.py:28]: Also, we might want to add a test that specifically verifies a `bool` is converted to an `int`? Like True -> 1 and False -> 0 to cover `and not (schema_type == "int" and isinstance(value, bool))` in the code.

File: python-avd/tests/pyavd/schema/test_avddataconverter.py
Comment (ClausHolbechArista) [reply python-avd/tests/pyavd/schema/test_avddataconverter.py:1]: I believe that is tested already when we have invalid data types.

File: python-avd/pyavd/_schema/avdschema.py [hunk 3]
```diff
@@ -163,6 +159,8 @@ def get_default_value(self, datapath: list) -> Any:
         Raises:
         -------
           AvdSchemaError if no default value is defined.
+
+        TODO: Remove this method since it seems like it is not used any longer.
```
Comment (ankudinov) [inline python-avd/pyavd/_schema/avdschema.py:163]: @ClausHolbechArista a very minor comment - is there anything blocking us from removing the method in this PR instead of adding TODO?

File: python-avd/tests/pyavd/schema/test_avddataconverter.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    avdschema = AvdSchema(schema=schema)
+    data = {"a": True, "b": True}
+    warnings = list(avdschema.convert(data))
+    assert len(warnings) == 1
+    assert str(warnings[0]) == (
+        "The input data model 'a' is deprecated and cannot be used in conjunction with the new data model 'b'. "
+        "This usually happens when a data model has been updated and custom structured configuration still uses the old model. "
+        "See arista.com for details."
```
Comment (ankudinov) [inline python-avd/tests/pyavd/schema/test_avddataconverter.py]: why arista.com and not avd.arista.com?

File: python-avd/tests/pyavd/schema/test_avddataconverter.py (hunk 1 above)
Comment (ankudinov) [reply python-avd/tests/pyavd/schema/test_avddataconverter.py:28]: It look like this PR already has a test for lowercase conversion (not sure if it was added after the comment).
But we certainly should add corner cases for boolean as Carl suggested.
I'd also add tests for all Ansible corner cases. Clear, it's Python and it might be converted beforehand in Ansible environments, but I'd add some test anyway for the following cases:

```yaml
port_channel:
  mode: on
string: "just a string"
integer: 1234
and_that_is_an_integer_too: 0xABCD
float: 12.34
version: "1.0" # is a string
boolean: true
# that's super weird, don't do that
but_that_is_a_string: !!str True
# there is a special `null` value for this case
and_this_is_not_empty:
a_better_null: ~
```
That can be another PR and likely we can skip something like null, but I'd certainly test for 0xABCD, empty value, float, etc.

**Review by ankudinov — APPROVED** [review]
Love this PR! Comments can be addressed in a different PR. LGTM.

File: python-avd/pyavd/_schema/avdschema.py (hunk 3 above)
Comment (ClausHolbechArista) [reply python-avd/pyavd/_schema/avdschema.py:163]: this is a test pr. so just that.

File: python-avd/tests/pyavd/schema/test_avddataconverter.py (hunk 4 above)
Comment (ClausHolbechArista) [reply python-avd/tests/pyavd/schema/test_avddataconverter.py]: typo I guess. I can try to find the place and fix.

Comment (ClausHolbechArista) [reply python-avd/tests/pyavd/schema/test_avddataconverter.py]: fixed

File: python-avd/tests/pyavd/schema/test_avddataconverter.py (hunk 2 above)
Comment (ClausHolbechArista) [reply python-avd/tests/pyavd/schema/test_avddataconverter.py:115]: added below. We don't really use it in the error, but it is used when emitting Ansible's variants.

File: python-avd/tests/pyavd/schema/test_avddataconverter.py (hunk 1 above)
Comment (ClausHolbechArista) [reply python-avd/tests/pyavd/schema/test_avddataconverter.py:28]: We are already testing all the regular features in the `test_avdschema_<bool|dict|int|list|str>.py` files.
The things you propose here Petr is more about deserializing YAML, and we will not test that here. We do test all those data types in the mentioned files.

**Comment by gmuloc** [comment]
For the record coverage in main:

```
pyavd/_schema/__init__.py                                                                                       0      0      0      0   100%
pyavd/_schema/avddataconverter.py                                                                              82     18     54      1    79%   94, 170-173, 191-221
pyavd/_schema/avdschema.py                                                                                     57     17     22      5    67%   60-61, 71-73, 128-129, 141-142, 147-148, 154-155, 167-170
pyavd/_schema/avdvalidator.py                                                                                 146      3    106      3    98%   170, 177, 251->exit, 275
pyavd/_schema/coerce_type.py                                                                                   27      9     12      1    69%   39-40, 46-50, 58-61
pyavd/_schema/constants.py                                                                                     11      0      0      0   100%
pyavd/_schema/models/__init__.py                                                                                0      0      0      0   100%
pyavd/_schema/models/avd_base.py                                                                               45      0      0      0   100%
pyavd/_schema/models/avd_indexed_list.py                                                                      152     15     48      8    88%   56-57, 203-204, 228-229, 253-254, 258, 262-263, 268-269, 284-285
pyavd/_schema/models/avd_list.py                                                                              135     14     46      8    87%   49-50, 54, 95, 101, 157-158, 190-191, 215->exit, 228-229, 237-238, 255
pyavd/_schema/models/avd_model.py                                                                             232     21    100     11    90%   64-65, 82-83, 234-235, 288-289, 319-321, 340-341, 379, 382, 401-402, 409-410, 421-422
pyavd/_schema/models/constants.py                                                                               5      0      0      0   100%
pyavd/_schema/models/eos_cli_config_gen_root_model.py                                                           7      0      0      0   100%
pyavd/_schema/models/eos_designs_root_model.py                                                                 55      4     22      4    90%   53-54, 57->60, 89, 94
pyavd/_schema/models/type_vars.py                                                                               9      0      0      0   100%
pyavd/_schema/store.py                                                                                         13      1      4      1    88%   14
pyavd/_schema/utils.py                                                                                          7      0      2      0   100%
```

in this PR:

```
pyavd/_schema/__init__.py                                                                                       0      0      0      0   100%
pyavd/_schema/avddataconverter.py                                                                              82      0     54      0   100%
pyavd/_schema/avdschema.py                                                                                     53      4     22      0    92%   165-168
pyavd/_schema/avdvalidator.py                                                                                 146      1    106      1    99%   170
pyavd/_schema/coerce_type.py                                                                                   27      9     12      1    69%   39-40, 46-50, 58-61
pyavd/_schema/constants.py                                                                                     11      0      0      0   100%
pyavd/_schema/models/__init__.py                                                                                0      0      0      0   100%
pyavd/_schema/models/avd_base.py                                                                               45      0      0      0   100%
pyavd/_schema/models/avd_indexed_list.py                                                                      152     15     48      8    88%   56-57, 203-204, 228-229, 253-254, 258, 262-263, 268-269, 284-285
pyavd/_schema/models/avd_list.py                                                                              135     14     46      8    87%   49-50, 54, 95, 101, 157-158, 190-191, 215->exit, 228-229, 237-238, 255
pyavd/_schema/models/avd_model.py                                                                             232     21    100     11    90%   64-65, 82-83, 234-235, 288-289, 319-321, 340-341, 379, 382, 401-402, 409-410, 421-422
pyavd/_schema/models/constants.py                                                                               5      0      0      0   100%
pyavd/_schema/models/eos_cli_config_gen_root_model.py                                                           7      0      0      0   100%
pyavd/_schema/models/eos_designs_root_model.py                                                                 55      4     22      4    90%   53-54, 57->60, 89, 94
pyavd/_schema/models/type_vars.py                                                                               9      0      0      0   100%
pyavd/_schema/store.py                                                                                         13      1      4      1    88%   14
pyavd/_schema/utils.py                                                                                          7      0      2      0   100%
```

**Review by gmuloc — APPROVED** [review]

**Review by carl-baillargeon — APPROVED** [review]
